Reviewed `0b3e7639` against `feature/registry-driven-rust-assembly`. Step 4a of #581.

## Reproduced

Workspace tests, strict Clippy, `cargo fmt --check`,
`RUSTDOCFLAGS='-D warnings' cargo doc --workspace --no-deps`. The committed
generated files are unchanged, the C smoke test is clean under
ASan/LSan/UBSan, and `covertest-kotlin` runs its 61 sections.

The five artifact kinds are the right cut, and `CEnum` retaining
`EnumValue` rather than a resolved number is the right call for the reason
the doc comment gives.

Four things below. The first is a hole in the new test, the second is a gap in
what the change is checked by, the third is a lost diagnostic, and the fourth
is older than this PR and is flagged because `CDomainConstant` now carries it
one layer further.

## The new test has one escape

`planned_artifacts_reach_the_file_through_the_frozen_plan` does bite: giving
`CPlanned` a `RefCell` field fails it with

```
a placed artifact retains RefCell, so rendering could resume planning
```

But one of the four forbidden strings is `"Registry,"`, with the comma, and
the field list is compared after whitespace is stripped. A `Registry` field
written last without a trailing comma therefore passes:

```rust
pub(crate) struct CPlanned {
    generation: Rc<GenerationPlan<crate::compile::CRepresentation>>,
    id: prebindgen_registry::generation::ArtifactId,
    registry: *const Registry
}
```

That is the one field the test most wants to forbid, and the test is green on
it. `cargo fmt` would restore the comma, but a test that only holds because a
separate gate runs is not holding. Stripping `prebindgen_registry` from the
field string and then matching bare `Registry` closes the escape, and keeps
the reason the comma was there in the first place.

## The value-opaque placement is not checked by anything

Reordering the enum mirrors ahead of the data-struct mirrors moves all five
committed generated files by 100 lines, so the regeneration check does pin
placement. It does not pin every section. `example-cbindgen` plans no
value-opaque at all — instrumenting the `planned` closure reports
`c-opaque-handle 1`, `c-value-opaque 0`, `c-tagged-union 2`, `c-callback 4` —
so swapping the value-opaque placement past the data-struct mirrors leaves
every generated file byte-identical and all 110 `prebindgen-c` tests green.

Of the eight sections this PR places, the value-opaque is the one nothing
pins. Declaring one value-opaque in the example would pin it, and would also
give the `c-value-opaque` rendering path its first end-to-end exercise.

## A duplicate identity now drops silently

`AssemblyBuilder::artifact` replaces a stored artifact only when the stored one
is unreachable, so two reachable artifacts sharing a key keep the first and
discard the second. Forcing `CDataStruct::key` to a constant shows what a
reader would see:

```
error[E0425]: cannot find type `foo_t` in this scope
    --> examples/example-cbindgen/generated/example_flat_x86_64.rs:1472:16
```

Before this PR the same collision produced a duplicate definition naming
`foo_t` twice. Both are errors in the generated crate rather than a prebindgen
diagnostic, so nothing is newly broken — but a missing type is the harder of
the two to trace back to two declared structs whose short names mangle alike.

A narrow guard is available and costs nothing. Instrumenting the two examples
to report every reachable duplicate gives only `ArtifactKey::Operation` —
converters legitimately reached from several sites — and never
`ArtifactKey::Artifact`. Asserting on a repeated reachable
`ArtifactKey::Artifact` is accordingly green across the workspace tests, both
examples and the regeneration check:

```rust
if let ArtifactKey::Artifact(id) = artifact.key() {
    assert!(
        !(self.artifacts[position].reachable() && artifact.reachable()),
        "two reachable artifacts share the identity {id}"
    );
}
```

That is `AssemblyBuilder`'s to add rather than this PR's, since every adapter
that names its own artifacts has the same exposure. Raising it here because
this PR is what gives the C adapter five kinds that name themselves.

## The converter filter now ends in a wildcard

`CbindgenBuilder::converters` matched all three variants before and now reads
`_ => None`. With eight variants the wildcard is the shorter code, but the
match is a filter over artifact kinds, so the next kind that should be visible
to a converter caller will be filtered out with nothing to say so. Listing the
five new variants next to `Wrapper` and `Const` keeps the compiler as the
reminder.

## `RepresentationDomain` decides meaning by spelling

This is not introduced here. `CDomainConstant` stores `ty: syn::Type` taken
from `domain.ty()`, so the spelling now survives into a planned artifact,
which is why it is worth naming now.

`RepresentationDomain` keeps its scalar as a `syn::Type` and recovers the kind
by matching the rendered token string. It does so twice: in `contains_expr`,
to decide whether a bounds check needs a NaN test, and in `niche_values`, to
pick the integer or float candidate generator, ending `_ => unreachable!()`.
The kind is not missing. `DomainScalar` produces it as `ScalarValue::$v` in
the same macro arm that produces the spelling, and throws it away.

A third site is in JniGen, which compares the domain's spelling against the
representation's, and it is the one that shows the consequence is not
theoretical. Spelling one integer `::core::primitive::u64` instead of `u64` —
the same type — fails the build with

```
convert!(Duration): domain type core_primitive_u64 does not match input
representation u64
```

So three sites decide equality of meaning by equality of text, and the one
that reports it is in the adapter this PR does not touch. `DomainScalar` is
publicly re-exported from `prebindgen_registry` and is not sealed, so a
downstream implementation choosing any other spelling of the same scalar
reaches the `unreachable!()`.

The fix is to carry the tag `DomainScalar` already computes rather than
re-deriving it from text. Flat's `ScalarKind` is not a drop-in — it has
`Bool`, `Isize` and `Usize`, and lacks `I128` and `U128` — so this needs
either a domain-side kind or a widened `ScalarKind`. Either way both string
matches and the `unreachable!()` go.

Worth its own issue rather than this PR. Nothing here blocks #587.

— Claude Code (Opus 5)

--- a/prebindgen-c/src/lib.rs
+++ b/prebindgen-c/src/lib.rs
@@ -471,7 +471,13 @@ impl CbindgenBuilder {
             .artifacts()
             .filter_map(|artifact| match artifact {
                 assembly::CFinalArtifact::Converter(converter) => Some(&**converter),
-                _ => None,
+                assembly::CFinalArtifact::Wrapper(_)
+                | assembly::CFinalArtifact::Const(_)
+                | assembly::CFinalArtifact::Memory(_)
+                | assembly::CFinalArtifact::DataStruct(_)
+                | assembly::CFinalArtifact::Enum(_)
+                | assembly::CFinalArtifact::DomainConstant(_)
+                | assembly::CFinalArtifact::Planned(_) => None,
             })
     }
 }

--- a/prebindgen-c/src/tests/boundary_invariants.rs
+++ b/prebindgen-c/src/tests/boundary_invariants.rs
@@ -473,7 +473,7 @@ fn planned_artifacts_reach_the_file_through_the_frozen_plan() {
         })
         .expect("the planned-artifact placement");
     let fields: String = fields.split_whitespace().collect();
-    for forbidden in ["CbindgenBuilder", "Registry,", "Compiled", "RefCell"] {
+    for forbidden in ["CbindgenBuilder", "Registry", "Compiled", "RefCell"] {
         assert!(
             !fields.contains(forbidden),
             "a placed artifact retains {forbidden}, so rendering could resume planning"

